libs/MeshKernel/src/CurvilinearGrid/CurvilinearGrid.cpp

```diff
@@ -146,120 +158,114 @@ void CurvilinearGrid::Delete(std::shared_ptr<Polygons> polygons, UInt polygonInd
-            gridIndices[ind] = {n, m};
-            ind++;
+            const auto faceCenters = ComputeFaceCenters();
+            m_RTrees.at(Location::Faces)->BuildTree(faceCenters, boundingBox);
+            m_facesRTreeRequiresUpdate = false;
+            m_boundingBoxCache = boundingBox;
         }
-    }
-
-    ind = 0;
-    for (UInt n = 0; n < NumN() - 1; n++)
-    {
-        for (UInt m = 0; m < NumM(); m++)
+        break;
+    case Location::Nodes:
+        if (m_nodesRTreeRequiresUpdate || m_boundingBoxCache != boundingBox)
         {
-
-            edges[ind].first = nodeIndices(n, m);
-            edges[ind].second = nodeIndices(n + 1, m);
-            ind++;
+            const auto nodes = ComputeNodes();
+            m_RTrees.at(Location::Nodes)->BuildTree(nodes, boundingBox);
+            m_nodesRTreeRequiresUpdate = false;
+            m_boundingBoxCache = boundingBox;
         }
-    }
-
-    for (UInt n = 0; n < NumN(); n++)
-    {
-        for (UInt m = 0; m < NumM() - 1; m++)
+        break;
+    case Location::Edges:
+        if (m_edgesRTreeRequiresUpdate || m_boundingBoxCache != boundingBox)
         {
-            edges[ind].first = nodeIndices(n, m);
-            edges[ind].second = nodeIndices(n, m + 1);
-            ind++;
+            m_edges = ComputeEdges();
+            const auto edgeCenters = ComputeEdgesCenters();
+            m_RTrees.at(Location::Edges)->BuildTree(edgeCenters, boundingBox);
+            m_edgesRTreeRequiresUpdate = false;
+            m_boundingBoxCache = boundingBox;
         }
+        break;
+    case Location::Unknown:
+    default:
+        throw std::runtime_error("Invalid location");
     }
-    edges.resize(ind);
-
-    return {nodes, edges, gridIndices};
 }
 
-bool CurvilinearGrid::IsValid() const
+meshkernel::UInt CurvilinearGrid::FindLocationIndex(Point point,
+                                                    Location location,
+                                                    const BoundingBox& boundingBox)
 {
-    if (m_gridNodes.size() == 0)
+    BuildTree(location, boundingBox);
+    const auto& rtree = m_RTrees.at(location);
+    if (rtree->Empty())
     {
-        return false;
-    }
-    if (NumM() < 2)
-    {
-        return false;
-    }
-    if (NumN() < 2)
-    {
-        return false;
+        throw AlgorithmError("Empty RTree");
     }
 
-    return true;
-}
+    rtree->SearchNearestPoint(point);
 
-CurvilinearGridNodeIndices CurvilinearGrid::GetNodeIndices(Point point)
-{
-    BuildTree(Location::Nodes);
-    SearchNearestLocation(point, Location::Nodes);
-    if (GetNumLocations(Location::Nodes) == 0)
+    if (rtree->GetQueryResultSize() <= 0)
     {
-        return {constants::missing::uintValue, constants::missing::uintValue};
+        throw AlgorithmError("Query result size <= 0.");
     }
 
-    const auto nodeIndex = GetLocationsIndices(0, Location::Nodes);
-    return m_gridIndices[nodeIndex];
+    return rtree->GetQueryResult(0);
```

Comment:
Should the missing::uIntValue be returned here rather than raise an exception?

Reply:
often the return value is used as an index, an exception is better